core/src/main/java/jenkins/telemetry/impl/StaplerDispatches.java

```diff
@@ -0,0 +1,119 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.telemetry.impl;
+
+import hudson.Extension;
+import hudson.PluginWrapper;
+import hudson.model.UsageStatistics;
+import hudson.util.VersionNumber;
+import jenkins.model.Jenkins;
+import jenkins.telemetry.Telemetry;
+import jenkins.util.SystemProperties;
+import net.sf.json.JSONObject;
+import org.kohsuke.MetaInfServices;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.EvaluationTrace;
+import org.kohsuke.stapler.StaplerRequest;
+
+import javax.annotation.Nonnull;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.time.LocalDate;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TimeZone;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+/**
+ * Telemetry implementation gathering information about Stapler dispatch routes.
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public class StaplerDispatches extends Telemetry {
+    @Nonnull
+    @Override
+    public LocalDate getStart() {
+        return LocalDate.of(2018, 10, 10);
+    }
+
+    @Nonnull
+    @Override
+    public LocalDate getEnd() {
+        return LocalDate.of(2019, 2, 1);
+    }
+
+    @Nonnull
+    @Override
+    public String getDisplayName() {
+        return "Stapler request handling";
+    }
+
+    @Nonnull
+    @Override
+    public JSONObject createContent() {
+        Map<String, Object> info = new TreeMap<>();
+        info.put("components", buildComponentInformation());
+        info.put("dispatches", buildDispatches());
+
+        return JSONObject.fromObject(info);
+    }
+
+    private Object buildDispatches() {
+        Set<String> currentTraces = traces;
+        traces = new TreeSet<>();
+        return currentTraces;
+    }
+
+    private Object buildComponentInformation() {
+        Map<String, String> components = new TreeMap<>();
+        VersionNumber core = Jenkins.getVersion();
+        components.put("jenkins-core", core == null ? "" : core.toString());
```

Comment:
How would it be possible for `Jenkins.getVersion` to respond with `null`? :confused:

Reply:
No idea, but it's `@CheckForNull` so I do that. The Javadoc of https://javadoc.jenkins.io/jenkins/model/Jenkins.html#getVersion-- mentions an example that does not apply, when you run Jenkins that way it's `2.146-SNAPSHOT` or similar. Might be a very rare occurrence such as running unit test in an IDE, but I choose to be careful here just in case.